src/Parts.elm

```diff
@@ -87,268 +135,222 @@ type alias Update' model msg =
-  c -> model
+type alias Accessors model container = 
+  { empty : model
+  , get : container -> model
+  , set : container -> model -> container
+  , map : (model -> model) -> container -> container
+  , reset : container -> container
+  }
 
 
-{-| Type of "setter": update component model `m` in c `c`. 
+{-| A collection abstracting over all the instances of a component.
+If you want to apply an action to all the instances of a component,
+use this data structure.
 -}
-type alias Set model c = 
-  model -> c -> c
+type alias Collection model container =
+  { empty : Indexed model
+  , get : container -> Indexed model
+  , set : container -> Indexed model -> container
+  , map : (Index -> model -> model) -> container -> container
+  , reset : container -> container
+  }
 
 
-{-| Lift a `view` to one which knows how to retrieve its `model` from 
-a c model `c`. 
--}
-embedView : Get model c -> View model a -> View c a
-embedView get view = 
-  get >> view 
 
+{-| Construct a collection:
 
-{-| Lift an `Update` from operating on `model` to a c model `c`. 
--}
-embedUpdate : 
-    Get model c
- -> Set model c
- -> Update model msg
- -> Update c msg
-embedUpdate get set update = 
-  \msg c -> 
-     update msg (get c) |> map1st (flip set c)
+    all : Collection Model (Container c)
+    all = 
+      collection .myField (\x y -> { y | myField = x }) 
 
-
-{-| Lift an explicit no-op `Update'` from operating on `model` to a c model `c`. 
 -}
-embedUpdate' :
-    Get model c
- -> Set model c
- -> Update' model msg
- -> Update' c msg
-embedUpdate' get set update = 
-  \msg c -> 
-    update msg (get c) 
-      |> map1st (Maybe.map (\x -> set x c))
-
-
-
--- INDEXED EMBEDDINGS
+collection
+  : (container -> (Indexed model))
+ -> (container -> (Indexed model) -> container)
+ -> Collection model container
+collection get set =
+  { empty = Dict.empty
+  , get = get
+  , set = set
+  , map = \f c -> get c |> Dict.map f |> set c 
+  , reset = flip set Dict.empty
+  }
 
- 
-{-| Type of indices. An index is a list of `Int` rather than just an `Int` to 
-support nested dynamically constructed elements: Use indices `[0]`, `[1]`, ...
-for statically known top-level components, then use `[0,0]`, `[0,1]`, ...
-for a dynamically generated list of components. 
--}
-type alias Index 
-  = List Int
- 
 
-{-| Indexed families of things.
--}
-type alias Indexed a = 
-  Dict Index a 
+{-| Generate accessors:
 
+    find = 
+      accessors all init
 
-{-| Fix a getter and setter for an `Indexed model` to a particular `Index`.
 -}
-indexed : 
-    Get (Indexed model) c
- -> Set (Indexed model) c
+accessors 
+  : Collection model container
  -> model
- -> (Index -> Get model c, Index -> Set model c)
-indexed get set model0 =  
-  ( \idx c -> Dict.get idx (get c) |> Maybe.withDefault model0
-  , \idx model c -> set (Dict.insert idx model (get c)) c
-  )
-  
-
--- EMBEDDING MESSAGES
-
+ -> Index
+ -> Accessors model container
+accessors collection model0 idx =
```

Comment:
It'd be neat if `accessors collection model0` had already computed `empty`, `get`, `set` etc. so that when you add the index, as little further computation as necessary is performed. Don't know if that's possible.